internal/worker/worker.go

```diff
@@ -0,0 +1,300 @@
+package worker
+
+import (
+	"archive/tar"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-containerregistry/pkg/crane"
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+)
+
+type Worker interface {
+	LoadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error
+	ReloadKmod(ctx context.Context, oldCfg, newCfg *kmmv1beta1.ModuleConfig) error
+	UnloadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error
+}
+
+type worker struct {
+	ip     ImagePuller
+	logger logr.Logger
+	mr     ModprobeRunner
+}
+
+func NewWorker(ip ImagePuller, mr ModprobeRunner, logger logr.Logger) Worker {
+	return &worker{
+		ip:     ip,
+		logger: logger,
+		mr:     mr,
+	}
+}
+
+func (w *worker) LoadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error {
+	imageName := cfg.ContainerImage
+
+	w.logger.Info("Pulling image", "name", imageName)
+
+	pr, err := w.ip.PullToDisk(ctx, imageName, cfg.InsecurePull)
+	if err != nil {
+		return fmt.Errorf("could not pull %q: %v", imageName, err)
+	}
+
+	if inTree := cfg.InTreeModuleToRemove; inTree != "" {
+		w.logger.Info("Unloading in-tree module", "name", inTree)
+
+		if err = w.mr.Run(ctx, "-rv", inTree); err != nil {
+			return fmt.Errorf("could not remove in-tree module %s: %v", inTree, err)
+		}
+	}
+
+	// TODO copy firmware
+	// TODO handle ModulesLoadingOrder
+
+	moduleName := cfg.Modprobe.ModuleName
+
+	var args []string
+
+	if cfg.Modprobe.RawArgs != nil {
+		args = cfg.Modprobe.RawArgs.Load
+	} else {
+		args = []string{"-vd", filepath.Join(pr.fsDir, cfg.Modprobe.DirName)}
+
+		if cfg.Modprobe.Args != nil {
+			args = append(args, cfg.Modprobe.Args.Load...)
+		}
+
+		args = append(args, moduleName)
+		args = append(args, cfg.Modprobe.Parameters...)
+	}
+
+	return w.mr.Run(ctx, args...)
+}
+
+func (w *worker) ReloadKmod(ctx context.Context, oldCfg, newCfg *kmmv1beta1.ModuleConfig) error {
+	return errors.New("not implemented")
```

Comment:
Should we add a TODO here?

Reply:
Added a TODO. But in fact, should we even implement this?
If the pod succeeds in unloading the old kmod, but fails to load the new one, we will not update the `NodeModulesConfig` to reflect that the old kmod is gone.
@yevgeny-shnaidman if we split reloads into two separate tasks and worker pods, we eliminate that problem.

Reply:
agree. we can remove this command and move this decision into the NMC controller